Hw16 classic web

demen777 opened the pull request:
CRUD приложение с Web UI и хранением данных в БД
Цель: Цель: разрабатывать полноценные классические Web-приложения Результат: Web-приложение полностью на стеке Spring
Необходимо:
1. Создать приложение с хранением сущностей в БД (можно взять библиотеку и DAO/репозитории из прошлых занятий)
2. Использовать классический View на Thymeleaf, classic Controllers.
3. Для книг (главной сущности) на UI должны быть доступные все CRUD операции. CRUD остальных сущностей - по желанию/необходимости.

demen777 committed:
First hw16_classic_web commit

demen777 committed:
Add BookControllerTest

demen777 committed:
Add /book/view tests into BookControllerTest

demen777 committed:
First implementation /book/edit

demen777 committed:
Use SpringBootTest in BookControllerTest

Commit:
Merge branch 'hw16_classic_web' of https://github.com/demen777/2019-11-otus-spring-denis-menschikov into hw16_classic_web

Commit:
Remove @ModelAttribute from editBookPost

Commit:
Add Dto for controller

Commit:
Add BookCommentDto to controller

demen777 committed:
Add /book/delete-comment handler

demen777 committed:
Add /book/delete handler

demen777 committed:
Add authors and genres view

Commit:
Add add_author.html and mappings in AuthorController
Add tests to AuthorControllerTest

Commit:
Add add_genre.html and mappings in GenreController
Add tests to GenreControllerTest

demen777 committed:
Merge branch 'master' into hw16_classic_web

stvort commented on hw16_classic_web/src/main/java/ru/otus/demen/books/controller/BookController.java:
```diff
@@ -0,0 +1,115 @@
+    private final AuthorDtoMapper authorDtoMapper;
+    private final GenreDtoMapper genreDtoMapper;
+    private final BookDtoMapper bookDtoMapper;
+    private final BookCommentDtoMapper bookCommentDtoMapper;
```
Ого) `@RequiredArgsConstructor` конечно спасает, но лучше сделать пару фасадов для сервисов и мапперов

demen777 replied:
Создал фасад для мапперов

stvort commented on hw16_classic_web/src/main/java/ru/otus/demen/books/controller/BookController.java:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    @GetMapping("/book/view")
+    public ModelAndView viewBookGet(@RequestParam("id") long id) {
```
Лучше чтобы названия отражали больше информации и при этом были глаголами. Например, `getBookDetailsView`, `insertBookComment` и т.д.)

demen777 replied:
Произвел изменения

stvort commented on hw16_classic_web/src/main/java/ru/otus/demen/books/controller/BookController.java:
```diff
@@ -0,0 +1,115 @@
+        return new RedirectView("/books");
+    }
+
+    @GetMapping("/book/delete-comment")
```
Так конечно удобно, но я бы все равно не делал удаление GET-ом. Он вроде как для получения данных

stvort commented on hw16_classic_web/src/test/java/ru/otus/demen/books/controller/AuthorControllerTest.java:
```diff
@@ -0,0 +1,91 @@
+
+@SpringBootTest(classes = ControllerTestConfiguration.class)
+@AutoConfigureMockMvc
+class AuthorControllerTest {
```
Лучше `@WebMvcTest`

demen777 replied:
Заменил на WebMvcTest.

stvort commented on hw16_classic_web/src/test/java/ru/otus/demen/books/controller/AuthorControllerTest.java:
```diff
@@ -0,0 +1,91 @@
+        when(authorService.getAll()).thenReturn(List.of(TOLSTOY));
+        ResultActions resultActions = mockMvc.perform(get("/authors"));
+        resultActions.andExpect(status().isOk())
+                .andExpect(content().contentType("text/html;charset=UTF-8"))
```
Еще имя view можно проверить и наличие атрибутов модели)

demen777 replied:
Добавлены проверки на имена view и на наличие атрибутов модели

Commit:
Create BookMappersFacade
Change controllers methods name
Use WebMvcTest instead SpringBootTest and AuthoConfigureMockMvc
Check in test view name and attributes in model
